### ISSUE TITLE
hdfe with tempvars

### ISSUE THREAD

**Issue by matthieugomez** [sergiocorreia/reghdfe#37]
I'm encountering issues when using `hdfe` inside a program

```
clear
set obs 3
gen a = 1
gen b = 1
gen x = 1
gen __000000 = 1
 hdfe x , a(a b) gen(new)
#__000000 already defined
 hdfe x , a(a b) gen(new)
# reghdfe error: data transformed with cache(save) requires cache(use)
```

**Comment by matthieugomez** [comment]
I also have an issue where hdfe returns "variable__000001 not found" 
happening in the line
```
keep `uid' `touse' `expandedvars' `weightvar' `absorb_keepvars' `cluster_keepvars' `tsvars ' `cachevars' `more_keepvars'
```
(where _000001 is an element of `more keepvars`)

**Comment by sergiocorreia** [comment]
Ok, so the first problem is because `__000000` is protected by Stata and should only be used together with tempvar. Example:

```stata


. clear

. set obs 3
obs was 0, now 3

. gen a = 1

. gen b = 1

. gen x = 1

. tempvar somevar

. gen `somevar' = 1

.  hdfe x , a(a b) gen(new)
(converged in 3 iterations)
```


Note that this works even if we use `__000000` instead of  ``somevar'` :

```stata

. clear

. set obs 3
obs was 0, now 3

. gen a = 1

. gen b = 1

. gen x = 1

. tempvar somevar

. gen __000000 = 1

.  hdfe x , a(a b) gen(new)
(converged in 3 iterations)
```

**Comment by sergiocorreia** [comment]
About the second case, is it possible that this is related with using __* variables without tempvars?

I'm trying to see how -tempvar- works, and from what I understand, it does not check if a variable exists, and just keeps an internal counter:

```stata
clear
gen __000000 = 2
tempvar foo
gen `foo' = 2
```

**Comment by matthieugomez** [comment]
Ok I do create the variable through tempvars though. 
I'll re look into it and get back to you.
Currently there's no way to assign variables created by hdfe to variable name created by tempvars right ? Maybe the problem comes from this.

**Comment by sergiocorreia** [comment]
That's true. On principle you could do replace and a loop to rename them to tempvars.

One problem with assigning variables to tempvars is that Stata would delete them when -hdfe- finishes running. The name of the variables needs to be given externally in order to avoid deletion

**Comment by matthieugomez** [comment]
I'm not sure we're talking about the same thing. When using hdfe inside a function, one generally wants to create variables with temporary names that don't overlap with existing variables. I think that
`hdfe x y, replace`
would be good syntax for this situation.

A typical use would be
`tempvar x1 x2
gen `x1' = x1
gen `x2' = x2
hdfe `x1' `x2', replace
```

**Comment by sergiocorreia** [comment]
> hdfe x y, replace could be good syntax for this situation.

Would this be equivalent to `hdfe x y, clear` but without dropping other variables or observations?

**Comment by matthieugomez** [comment]
Yes!

**Comment by sergiocorreia** [comment]
I'll add it as an issue..

**Comment by matthieugomez** [comment]
Ok I think I found a reproductible example 

```
sysuse nlsw88.dta, clear
fvrevar c.tenure#i.age
hdfe `=r(varlist)', a(race married) gen(new)
```

**Comment by sergiocorreia** [comment]
This seems to be caused by a very subtle bug:

reghdfe uses fvrevar internally to expand factor variables and then give them a human readable name
E.G: `reghdfe price i.foreign, a(turn) cache(save)`

Deep in the code, it depends on some chars created by fvrevar:

```stata
sysuse auto
fvrevar i.foreign
char list
```
And the bug lies in assuming that there were no previous fvrevars in the code.

A quick workaround that you can use for now would be:

```stata
fvrevar c.tenure#i.age

* new code
local vars = r(varlist)
foreach var of local vars {
	char `var'[fvrevar]
}

hdfe `vars', a(race married) gen(new)
```

A saner fix would be to do this within reghdfe, but since I'll probably ditch this approach in favor of using [fvstrip](https://github.com/markeschaffer/stata-utilities/blob/master/fvstrip.ado), it feels like too much hassle for a short-term fix. I'll leave the issue open to remind myself of this though.

**Comment by matthieugomez** [comment]
I don't mind the quick workaround. thanks for diving into your code!
